nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/CopyAzureBlobStorage_v12.java

```diff
@@ -0,0 +1,388 @@
+            blockBlobStageBlockFromUrlOptions.setSourceRange(new BlobRange(offset, count));
+
+            if (httpAuthorization != null) {
+                blockBlobStageBlockFromUrlOptions.setSourceAuthorization(httpAuthorization);
+            }
+            blockBlobStageBlockFromUrlOptions.setSourceRequestConditions(sourceRequestConditions);
+            final int statusCode = blockBlobClient.stageBlockFromUrlWithResponse(blockBlobStageBlockFromUrlOptions, null, Context.NONE).getStatusCode();
+            if (statusCode != 201) {
+                throw new ProcessException(String.format("Failed staging one or more blocks (status: %d)", statusCode));
+            }
+            blockIds.add(base64BlockId);
+            offset += count;
+            blockId++;
+        }
+
+        final BlockBlobCommitBlockListOptions options = new BlockBlobCommitBlockListOptions(blockIds);
+        options.setRequestConditions(destinationRequestConditions);
+        final Response<BlockBlobItem> response = blockBlobClient.commitBlockListWithResponse(options, null, Context.NONE);
+        final int statusCode = response.getStatusCode();
+        if (statusCode != 201) {
+            throw new ProcessException(String.format("Failed committing block list (status: %d)", statusCode));
+        }
+    }
+
+    private static HttpAuthorization getHttpAuthorization(final AzureStorageCredentialsDetails_v12 credentialsDetails) {
+        switch (credentialsDetails.getCredentialsType()) {
+            case ACCESS_TOKEN -> {
+                TokenCredential credential = tokenRequestContext -> Mono.just(credentialsDetails.getAccessToken());
+                return getHttpAuthorizationFromTokenCredential(credential);
+            }
+            case MANAGED_IDENTITY -> {
+                final ManagedIdentityCredential credential = new ManagedIdentityCredentialBuilder()
+                        .clientId(credentialsDetails.getManagedIdentityClientId())
+                        .build();
+                return getHttpAuthorizationFromTokenCredential(credential);
+            }
+            case SERVICE_PRINCIPAL -> {
+                final ClientSecretCredential credential = new ClientSecretCredentialBuilder().clientId(
+                                credentialsDetails.getServicePrincipalClientId()
+                        ).clientSecret(credentialsDetails.getServicePrincipalClientSecret())
+                        .tenantId(credentialsDetails.getServicePrincipalTenantId()).build();
+                return getHttpAuthorizationFromTokenCredential(credential);
+            }
+        }
+        return null;
+    }
+
+    private static HttpAuthorization getHttpAuthorizationFromTokenCredential(final TokenCredential credential) {
+        final TokenRequestContext tokenRequestContext = new TokenRequestContext();
+        tokenRequestContext.setScopes(Collections.singletonList("https://storage.azure.com/.default"));
```

Comment:
Recommend declaring a static variable for the default scope URL, something like `AZURE_STORAGE_DEFAULT_SCOPE`.

Reply:
Fixed in 89f3ddd3d31e6658cb84cd1684bec6114ebc508c.